modelmapper/loader.py

```diff
@@ -9,54 +10,131 @@ def select(fields):
+        query = select([table.c.id]).where(table.c.signature == signature)
+        query_result = session.execute(query)
+        result = query_result.fetchone()
+        result = None if result is None else result[0]
+        return result
+
+    def insert_chunk_of_data_to_db(self, session, model, chunk):
+        """Add row signature to row then run Base class logic"""
+        new_chunk = self.add_row_signature(chunk)
+        return super().insert_chunk_of_data_to_db(session, model, new_chunk)
+
+
+class UniqueSignatureSqlalchemyLoaderMixin(SignatureSqlalchemyMixin):
+    """
+    Sqlalchemy Specific Loader with Signatures that DOES NOT ALLOW DUPLICATES.
+    It appends a signature (64-bit murmur hash of the row) and only uploads if
+    it is not already in the table.
     """
-    Postgres Specific Loader that backs up raw data into S3.
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.all_recent_rows_signatures = set()
+
+    def add_row_signature(self, chunk):
+        """Generate hash of row, yield row only if it has not been generated recently"""
+        chunk_gen = super().add_row_signature(chunk)
+        for row in chunk_gen:
+            signature = row.get('signature')
+            if signature and signature not in self.all_recent_rows_signatures:
+                self.all_recent_rows_signatures.add(signature)
+                yield row
+
+    def insert_row_into_db(self, row: dict, session, model):
+        """Insert row only if the signature is unique to the table"""
+        if row['signature'] and not self.get_id_by_signature(session, model, row['signature']):
+            row_obj = model(**row)
+            try:
+                session.add(row_obj)
+                session.flush()
+                return row_obj
+            except Exception:
+                self.logger.error(f"Error on inserting row of data: {row}")
+                raise
+
+
+class SqlalchemySnapshotLoaderMixin(SignatureSqlalchemyMixin):
+    """
+    Sqlalchemy Specific Loader with Snapshot Auxilary Table for row dupes.
     It adds the records to the snapshot model.
     """
     SNAPSHOT_MODEL = None
 
-    def backup_data(self, content, key, metadata):
-        self.put_file_on_s3(content=content, key=key, metadata=metadata)
-
-    def insert_chunk_of_data_to_db(self, session, table, chunk):
+    def insert_chunk_of_data_to_db(self, session, model, chunk):
+        table = model.__table__
         snapshot_table = self.SNAPSHOT_MODEL.__table__
         new_chunk = self.add_row_signature(chunk) if self.settings.ignore_duplicate_rows_when_importing else chunk  # NOQA
         count = 0
         if new_chunk:
             for row in new_chunk:
                 id_ = None
                 if row['signature']:
-                    id_ = get_id_by_signature(session, table, row['signature'])
+                    id_ = self.get_id_by_signature(session, table, row['signature'])
```

Comment:
Same here, why can't we let Postgres handle signature uniqueness.

Reply:
I am will leave this the same because it is for different worker that may require the check for the integrity of their table